### PR TITLE
Energy Guns

### PR DESCRIPTION
**Pull request by FluffiestFloof**
## About the PR

The Head of Security special energy gun and the normal energy gun from TG.

## Why / Balance

The Head of Security deserves funny toys and this one has _two_ firemode, using the same amount of shots as the TG version.
It can shoot disabler bolts and funny projectile lasers specifically made for it because hitscan lasers is sad.

It spawns inside their locker and is also a steal objective similarly to TG. Good luck for that one.

## Technical details

For some unknown magical reason, I couldn't place my new projectiles into the DeltaV folders, it would just always error out and only work if I placed them directly in the base game files.
Why? _shrug_

~~The other big note is that while we have firemodes (provided by the CHIMP gun out of all places) it doesn't really change the sprite so it gets to stay red. The rest of the sprites are still there, if they ever get used, but at that point it should just get added to upstream properly alongside the energy guns.~~

A new system and component was made for the EnergyGun that allows for it to change between lethal and disable firemode while also changing sprites. It's the stunbaton and firemode system duct taped together.

## Media
Pew pew it has a projectile, can change firemode and sounds different

https://github.com/DeltaV-Station/Delta-v/assets/9823203/537e7387-d45e-4747-bd47-bea59f7b6db0

It now can change sprite so cool

https://github.com/DeltaV-Station/Delta-v/assets/9823203/22affbdf-3f91-42b9-929f-223add325e5b

The examine menu
ss+(2023-11-03+at+10 37 46)

## Breaking changes

**Changelog**

:cl:
- add: Added the X-01 MultiPhase Energy Gun for the HoS.

### CODE REVIEW

**Comment by Carolyn3114** [comment]
this is awesome, I love how it's the first projectile laser gun, makes it feal real unique. Perhaps the disabler shot could also go through walls?

**Comment by ps3moira** [comment]
Finally! HoS has their own gun now :)

**Comment by Adrian16199** [comment]
Is it me or did the laser loose 10 magazine shots mid firing.

**Comment by ps3moira** [comment]
> Is it me or did the laser loose 10 magazine shots mid firing.

Pretty sure that's how it is in SS13, lethal shots cost 10 laser energy. Prevents people from spam swapping firing modes.

**Comment by FluffiestFloof** [comment]
> Is it me or did the laser loose 10 magazine shots mid firing.

It's because when you change firemode and start shooting it updates the "total amount of shots you can have with this firemode"
If you stay with the disable mode it has 20 shots, if you stay with the lethal mode it has 10 shots.
So you could shoot 5 lethal then 10 disabler

**Comment by ps3moira** [comment]
I think you could copy the code from Stun baton when swapping sprites for disable and lethal mode

**Comment by DebugOk** [comment]
Funni big update

**Comment by Tryded** [comment]
We should probably add energy guns with this PR since it adds the code to do it

**Comment by FluffiestFloof** [comment]
> We should probably add energy guns with this PR since it adds the code to do it

The main question is what will we do with them? At best Captain could get it but they already have their laser pistol and we don't want them to be roundstart nor give it to HoP from what was discussed in discord .

File: Content.Server/DeltaV/Weapons/Ranged/Components/EnergyGunComponent.cs [hunk 1]
```diff
@@ -0,0 +1,52 @@
+using Robust.Shared.Prototypes;
+using Robust.Shared.Serialization.TypeSerializers.Implementations.Custom.Prototype;
+using Content.Server.DeltaV.Weapons.Ranged.Systems;
+
+namespace Content.Server.DeltaV.Weapons.Ranged.Components;
+
+/// <summary>
+/// DeltaV Addition: Allows for energy gun to switch between lethal and disable. This also changes sprites accordingly.
+/// Yes this is a mashup of the StunbatonSystem and BatteryWeaponFireModesSystem
+/// </summary>
```
Comment (DEATHB4DEFEAT) [inline Content.Server/DeltaV/Weapons/Ranged/Components/EnergyGunComponent.cs]:
```suggestion
/// <summary>
/// Allows for energy gun to switch between lethal and disable. This also changes sprites accordingly.
/// </summary>
/// <remarks>Yes this is a mashup of the StunbatonSystem and BatteryWeaponFireModesSystem</remarks>
```

File: Content.Server/DeltaV/Weapons/Ranged/Systems/EnergyGunSystem.cs
Comment (DEATHB4DEFEAT) [inline Content.Server/DeltaV/Weapons/Ranged/Systems/EnergyGunSystem.cs:1]: This should be made into a system that is usable for more than just "guns that can either disable or kill at a moment's notice"

File: Resources/Audio/DeltaV/license.txt [hunk 2]
```diff
@@ -2,3 +2,4 @@ typewriter.ogg is modified from https://www.freesoundslibrary.com/typewriter-sou
 axiom.ogg is modified from free sound effects from https://www.fesliyanstudios.com which are licensed under the CC0 1.0 License. This modified file follows the same license.
 jackdaw.ogg is modified from atreides.ogg which was taken from https://gitlab.com/cmdevs/colonial-warfare/-/tree/32cb5892413243cc74bb2d11df8e3085f8ef1164/sound/weapons which is licensed under the CC-BY-SA 3.0 License. This modified file follows the same license.
 universal.ogg is modified from free sound effects from https://www.fesliyanstudios.com which are licensed under the CC0 1.0 License. This modified file follows the same license.
+dry_fire.ogg and laser.ogg both from tgstation at commit https://github.com/tgstation/tgstation/commit/2532911353d63661b735004f2895103d45858b50
```
Comment (DEATHB4DEFEAT) [inline Resources/Audio/DeltaV/license.txt]: Link to the files in the tree, commit unrelated.

File: Resources/Prototypes/DeltaV/Objectives/traitor.yml [hunk 3]
```diff
@@ -19,3 +19,17 @@
...
     # owner: job-name-hop
+
+- type: entity # Head of Security steal objective.
+  noSpawn: true
+  parent: BaseTraitorStealObjective
+  id: HoSGunStealObjective
+  components:
+  - type: Objective
+    # hos will mostly be using that gun to stop you from stealing it
+    difficulty: 3
```
Comment (DEATHB4DEFEAT) [inline Resources/Prototypes/DeltaV/Objectives/traitor.yml]:
```suggestion
    difficulty: 3 # HoS will mostly be using the gun to stop you from stealing it
```

File: Resources/Prototypes/Entities/Objects/Weapons/Guns/Projectiles/projectiles.yml [hunk 4]
```diff
@@ -791,3 +791,48 @@
             - Impassable
             - HighImpassable
     - type: GrapplingProjectile
+
+# DeltaV projectiles
+# For whatever magical reasons, I cannot put those in the DeltaV folders without having errors.
```
Comment (DEATHB4DEFEAT) [inline Resources/Prototypes/Entities/Objects/Weapons/Guns/Projectiles/projectiles.yml:796]: Weird

File: Resources/Textures/DeltaV/Objects/Weapons/Guns/Battery/multiphase_energygun.rsi/meta.json [hunk 5]
```diff
@@ -0,0 +1,64 @@
+{
+  "version": 1,
+  "license": "CC-BY-SA-3.0",
+  "copyright": "taken from tg station at commit https://github.com/tgstation/tgstation/commit/8b7f8ba6a3327c7381967c550f185dffafd11a57",
```
Comment (DEATHB4DEFEAT) [inline Resources/Textures/DeltaV/Objects/Weapons/Guns/Battery/multiphase_energygun.rsi/meta.json]: Link to file(s) in tree, commit seems unrelated.

File: Resources/Textures/DeltaV/Objects/Weapons/Guns/Battery/multiphase_energygun.rsi/meta.json [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+			"name": "mode_lethal"
+		},
+		{
+			"name": "mag-unshaded-0",
+            "delays": [
+                [
+                    0.3,
+                    0.3
+                ]
+            ]
```
Comment (DEATHB4DEFEAT) [inline Resources/Textures/DeltaV/Objects/Weapons/Guns/Battery/multiphase_energygun.rsi/meta.json]:
```suggestion
            "delays": [[ 0.3, 0.3 ]]
```

**Review by DEATHB4DEFEAT — DISMISSED** [review]

**Comment by Tryded** [comment]
> > We should probably add energy guns with this PR since it adds the code to do it
> 
> The main question is what will we do with them? At best Captain could get it but they already have their laser pistol and we don't want them to be roundstart nor give it to HoP from what was discussed in discord .

We could make it orderable from cargo and researchable

File: Content.Server/DeltaV/Weapons/Ranged/Systems/EnergyGunSystem.cs
Comment (FluffiestFloof) [reply Content.Server/DeltaV/Weapons/Ranged/Systems/EnergyGunSystem.cs:1]: Yeah that would be a good idea, sadly that would require way more code thing that I'm not smart enough to do.

**Comment by Carolyn3114** [comment]
vid 4 normal energy gun where

**Comment by Tryded** [comment]
> > We should probably add energy guns with this PR since it adds the code to do it
> 
> The main question is what will we do with them? At best Captain could get it but they already have their laser pistol and we don't want them to be roundstart nor give it to HoP from what was discussed in discord .

Also, when you say roundstart, do you mean it spawns in their bag? The normal energy gun should take up as much space as an laser carbine and some should spawn in the armory if you meant what I said.

**Comment by Tryded** [comment]
Also to compensate with lasers being worse than energy guns, lasers should get 19 rounds of fire instead of 16 + Energy guns should have 16 kill shots and 20 disable if possible

**Review by DebugOk — APPROVED** [review]
I'm sure it'll be fine.

It should still be made into a more generic system imo, but this'll do

**Comment by DebugOk** [comment]
Fuck it, we ball

**Pull request merged**